src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -175,223 +138,310 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 return;
             }
 
-            var declarationStatement = (StatementSyntax)declarator.Parent.Parent;
-
-            // If we convert this to 'if (o is Type x)' then 'x' will not be definitely assigned
-            // in the Else branch of the IfStatement, or after the IfStatement. Make sure
-            // that doesn't cause definite assignment issues.
-            if (!CheckDefiniteAssignment(semanticModel, localSymbol,
-                    declarationStatement, targetStatement, parentBlock, isNegativeNullCheck))
+            if (!CanSafelyConvertToPatternMatching())
             {
                 return;
             }
 
             // Looks good!
             var additionalLocations = ImmutableArray.Create(
-                declarationStatement.GetLocation(),
-                targetStatement.GetLocation(),
-                leftmostCondition.GetLocation(),
+                localStatement.GetLocation(),
+                comparison.GetAncestor<StatementSyntax>().GetLocation(),
+                comparison.GetLocation(),
                 asExpression.GetLocation());
 
             // Put a diagnostic with the appropriate severity on the declaration-statement itself.
             syntaxContext.ReportDiagnostic(Diagnostic.Create(
                 GetDescriptorWithSeverity(styleOption.Notification.Value),
-                declarationStatement.GetLocation(),
+                localStatement.GetLocation(),
                 additionalLocations));
-        }
-
-        // Ensure that all usages of the pattern variable are
-        // in scope and definitely assigned after replacement.
-        private static bool CheckDefiniteAssignment(
-            SemanticModel semanticModel,
-            ISymbol localVariable,
-            StatementSyntax declarationStatement,
-            StatementSyntax targetStatement,
-            BlockSyntax parentBlock,
-            bool isNegativeNullCheck)
-        {
-            var statements = parentBlock.Statements;
-            var targetIndex = statements.IndexOf(targetStatement);
-            var declarationIndex = statements.IndexOf(declarationStatement);
 
-            // Since we're going to remove this declaration-statement,
-            // we need to first ensure that it's not used up to the target statement.
-            if (declarationIndex + 1 < targetIndex)
+            bool CanSafelyConvertToPatternMatching()
             {
-                var dataFlow = semanticModel.AnalyzeDataFlow(
-                    statements[declarationIndex + 1], 
-                    statements[targetIndex - 1]);
-
-                if (dataFlow.ReadInside.Contains(localVariable) || 
-                    dataFlow.WrittenInside.Contains(localVariable))
+                bool DataFlowsIn(SyntaxNode statementOrExpression)
```

Comment:
having multiple levels of local function nesting is super hard to read and to understand the actual flow of logic.  can we not overdo it here?

Reply:
this name is pretty vague. without reading the code i had no idea it related to the local symbol we were looking at. sometimes passing things as parameters helps better understand what the code is intended to be doing.

Reply:
For future reference, I just use it for quick prototyping, so I don't have to add/remove parameters all along. I'll change it for the next iteration.